_posts/2025-02-17-using-langchain4j-to-analyze-pdf-documents.adoc

```diff
@@ -0,0 +1,183 @@
+---
+layout: post
+title: 'Using LangChain4j to analyze PDF documents'
+date: 2025-02-17
+tags: user-story langchain4j llm ai
+synopsis: 'Learn how to extract structured metadata from PDF documents using LangChain4j and AI to automate document analysis.'
+author: melloware
+thumbnailimage: /assets/images/posts/quarkus-user-stories/melloware/ksm-logo.png
+---
+
+:imagesdir: /assets/images/posts/quarkus-user-stories/melloware
+ifdef::env-github,env-browser,env-vscode[:imagesdir: ../assets/images/posts/quarkus-user-stories/melloware]
+
+In my consulting work, clients frequently present us with challenging problems that require innovative solutions.
+Recently, we were tasked with extracting structured metadata from PDF documents through automated analysis. Below, I'll share a simplified version of this real-world challenge and how we approached it.
+
+== Use Case
+
+Our client receives compressed archives (.zip files) containing up to hundreds of portable document format (PDF) lease documents that need review. Each document contains property lease details that must be validated for accuracy. The review process involves checking various business rules - for example, identifying leases with terms shorter than 2 years. Currently, this document validation is done manually, which is time-consuming. The client wants to automate and streamline this review workflow to improve efficiency.
+
+Some complications with these lease documents are:
+
+* The documents are not in a standard format so each lease may be written in a different way by a different property manager.
+* The documents may be scanned, so the text is sometimes human writing and not typewritten.
+* The documents may contain multiple pages, which are not always in the same order.
+* The lease terms may not be an actual date but written as "Expires five years from the start date" or "Expires on the anniversary of the start date".
+* Metadata such as acreage and tax parcel information is needed by our client to validate the lease details.
+
+You can understand why this is time consuming for a human to review and validate the documents.
+
+== Our Solution
+
+After consulting with https://github.com/dliubarskyi[Dmytro Liubarskyi] and collaborating with the Quarkus team, we implemented a solution using LangChain4j for document metadata extraction. We chose https://ai.google.dev/docs/gemini_api_overview[Google Gemini] as our Large Language Model (LLM) since it excels at PDF analysis through its built-in Optical Character Recognition (OCR) capabilities, enabling accurate text extraction from both digital and scanned documents.
+
+== Technical Details
+
+The application is built using:
+
+* Quarkus - A Kubernetes-native Java framework
+* LangChain4j - Java bindings for LangChain to interact with LLMs  
+* Google Gemini AI - For PDF document analysis and information extraction
+* RESTEasy - For handling multipart file uploads
```

Comment:
Let's please use the `Quarkus REST` terminology here instead of `RESTEasy`

Reply:
Switched to Quarkus REST